Show more details in logs when S3 storage error happens

## Situation

If there are S3 problems when fetching data from storage - e.g. to fetch sourcecode zip file - the 
log does only contain following information

Log warn entry: 
```
The execution of 'Read and copy storage: sourcecode.zip for job: ebbfddbe-e8d7-49cb-9518-ab39de35e635' failed with 'Was not able to fetch object from s3 bucket:sourcecode.zip'. Will try to handle it resilient
```
After this an error log entry is written
```
Execution of job uuid:ebbfddbe-e8d7-49cb-9518-ab39de35e635 failed
```
At stacktrace level we see only "Was not able to fetch object from s3 bucket:sourcecode.zip" at the end but not any details.
```java
java.io.IOException: Storage read failed. Even after 4 tempts it was not possible to execute 'Read and copy storage: sourcecode.zip for job: ebbfddbe-e8d7-49cb-9518-ab39de35e635'! Will throw last cause. 	
at  com.mercedesbenz.sechub.pds.job.PDSWorkspaceService$1.throwException(PDSWorkspaceService.java:115) 	
at  com.mercedesbenz.sechub.pds.util.PDSResilientRetryExecutor.execute(PDSResilientRetryExecutor.java:127) 	
at  com.mercedesbenz.sechub.pds.util.PDSResilientRetryExecutor.execute(PDSResilientRetryExecutor.java:74) 	
at  com.mercedesbenz.sechub.pds.job.PDSWorkspaceService.importWantedFilesFromJobStorage(PDSWorkspaceService.java:181) 	
at  com.mercedesbenz.sechub.pds.job.PDSWorkspaceService.prepare(PDSWorkspaceService.java:146) 	
at  com.mercedesbenz.sechub.pds.execution.PDSExecutionCallable.pepareWorkspace(PDSExecutionCallable.java:166) 	
at  com.mercedesbenz.sechub.pds.execution.PDSExecutionCallable.call(PDSExecutionCallable.java:104) 	
at  com.mercedesbenz.sechub.pds.execution.PDSExecutionCallable.call(PDSExecutionCallable.java:50) 	
at  java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) 	
at  java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) 	
at  java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) 	
at  java.base/java.lang.Thread.run(Thread.java:833) 
Caused by: java.io.IOException: Was not able to fetch object from s3 bucket:sourcecode.zip 	
at  com.mercedesbenz.sechub.storage.s3.AwsS3JobStorage.fetch(AwsS3JobStorage.java:134) 	
at  com.mercedesbenz.sechub.pds.job.PDSWorkspaceService.readAndCopyStorageToFileSystem(PDSWorkspaceService.java:204) 	
at  com.mercedesbenz.sechub.pds.job.PDSWorkspaceService.lambda$importWantedFilesFromJobStorage$1(PDSWorkspaceService.java:181) 	
at  com.mercedesbenz.sechub.pds.util.PDSResilientRetryExecutor.lambda$execute$0(PDSResilientRetryExecutor.java:71) 	
at  com.mercedesbenz.sechub.pds.util.PDSResilientRetryExecutor.execute(PDSResilientRetryExecutor.java:109) 	... 10 common frames omittedsechub.pds.util.PDSResilientRetryExecutor.execute(PDSResilientRetryExecutor.java:109) 	... 10 common frames omitted
```

## Wanted
As an admin I want to see more details about the S3 storage problem, so it's easier to identify the origin problem (e.g. network connection errors etc.)

## Solution
Improve exception handling inside `AwsS3JobStorage` : 

The details shall be visible inside the stacktrace

--- a/sechub-storage-s3-aws-test/README.md
+++ b/sechub-storage-s3-aws-test/README.md
@@ -10,6 +10,6 @@ from an IDE (e.g. Eclipse) because having multiple Spring Boot Applications insi
 `sechub-storage-s3-aws-test` is a dedicated gradle project which has only dependencies to storage
 parts (core, s3-aws) and nothing else. The only Spring Boot Application is `S3MockApplication`.
 
-With this project is possible to use S3Mock for Junit tests without special processes, docker
+With this project it is possible to use S3Mock for Junit tests without special processes, docker
 containers etc. but also without any side effects.
 

--- a/sechub-storage-s3-aws-test/README.md
+++ b/sechub-storage-s3-aws-test/README.md
@@ -1,21 +1,15 @@
 <!-- SPDX-License-Identifier: MIT --->
-## About sechub-shared-kernel-storage
-This is a special testproject having a dependency to s3mock.
 
-S3mock is itself a spring application. So integrating it for junit tests
-can have very much side effects when starting from an IDE (e.g. eclipse).
+## About sechub-storage-s3-aws-test
+This is a special test project having a dependency to S3mock.
 
-So, this is a dedicated gradle project which has only dependencies to storage
-parts (core, s3-aws) and nothing else.
+S3mock is itself a Spring application, which means that integrating it for "normal" junit tests
+for existing Spring Boot applications (like SecHub server) will lead to side effects when starting 
+from an IDE (e.g. Eclipse) because having multiple Spring Boot Applications inside classpath.
 
-#### What happens with s3mock otherwise
-`S3MockApplication` itself is a spring boot application and will have problems
-to be ready for junit test execution when we got spring stereotypes inside other
-code:
+`sechub-storage-s3-aws-test` is a dedicated gradle project which has only dependencies to storage
+parts (core, s3-aws) and nothing else. The only Spring Boot Application is `S3MockApplication`.
 
-- slow start
-- unexpected behaviours
-- ...
+With this project is possible to use S3Mock for Junit tests without special processes, docker
+containers etc. but also without any side effects.
 
-So we can simply use s3mocking for junit tests without special processes, docker
-containers etc.

--- a/sechub-storage-s3-aws-test/src/test/java/com/mercedesbenz/sechub/storage/s3/AwsS3JobStorageTest.java
+++ b/sechub-storage-s3-aws-test/src/test/java/com/mercedesbenz/sechub/storage/s3/AwsS3JobStorageTest.java
@@ -134,7 +134,7 @@ class AwsS3JobStorageTest {
         IOException exception = assertThrows(IOException.class, () -> jobStorage.store(fileName, inputStream, contentLength));
 
         /* test */
-        assertEquals("Store of " + fileName + " to s3 failed", exception.getMessage());
+        assertEquals("Store of: " + fileName + " to S3 bucket: " + bucketName + " failed", exception.getMessage());
         assertEquals(exception.getCause(), amazonClientException);
 
     }

--- a/sechub-storage-s3-aws/src/main/java/com/mercedesbenz/sechub/storage/s3/AwsS3JobStorage.java
+++ b/sechub-storage-s3-aws/src/main/java/com/mercedesbenz/sechub/storage/s3/AwsS3JobStorage.java
@@ -77,6 +77,14 @@ public class AwsS3JobStorage implements JobStorage {
     }
 
     @Override
+    public void store(String name, InputStream inputStream) throws IOException {
+        requireNonNull(name, "name may not be null!");
+        requireNonNull(inputStream, "inputStream may not be null!");
+
+        storeInS3(name, inputStream, null);
+    }
+
+    @Override
     public void store(String name, InputStream inputStream, long contentLength) throws IOException {
         requireNonNull(name, "name may not be null!");
         requireNonNull(inputStream, "inputStream may not be null!");
@@ -90,37 +98,36 @@ public class AwsS3JobStorage implements JobStorage {
 
     @Override
     public Set<String> listNames() throws IOException {
-        if (!client.doesBucketExistV2(bucketName)) {
-            return Collections.emptySet();
-        }
-        Set<String> set = new LinkedHashSet<>();
-        ObjectListing data = client.listObjects(bucketName, getObjectPrefix());
-
-        String prefix = getObjectPrefix();
-        int prefixLength = prefix.length();
-
-        List<S3ObjectSummary> summaries = data.getObjectSummaries();
-        for (S3ObjectSummary summary : summaries) {
-            if (summary == null) {
-                continue;
+        try {
+            if (!client.doesBucketExistV2(bucketName)) {
+                return Collections.emptySet();
             }
-            String key = summary.getKey();
-            if (key == null || key.length() <= prefixLength) {
-                continue;
+            Set<String> set = new LinkedHashSet<>();
+            ObjectListing data = client.listObjects(bucketName, getObjectPrefix());
+
+            String prefix = getObjectPrefix();
+            int prefixLength = prefix.length();
+
+            List<S3ObjectSummary> summaries = data.getObjectSummaries();
+            for (S3ObjectSummary summary : summaries) {
+                if (summary == null) {
+                    continue;
+                }
+
+                String key = summary.getKey();
+                if (key == null || key.length() <= prefixLength) {
+                    continue;
+                }
+
+                String filenameOnly = key.substring(prefixLength);
+
+                set.add(filenameOnly);
             }
-            String filenameOnly = key.substring(prefixLength);
+            return set;
 
-            set.add(filenameOnly);
+        } catch (Exception e) {
+            throw new IOException("Was not able to list names for bucket: " + bucketName, e);
         }
-        return set;
-    }
-
-    private String getObjectName(String name) {
-        return getObjectPrefix() + name;
-    }
-
-    private String getObjectPrefix() {
-        return storagePath + "/" + jobUUID + "/";
     }
 
     @Override
@@ -131,35 +138,42 @@ public class AwsS3JobStorage implements JobStorage {
 
             return client.getObject(bucketName, objectName).getObjectContent();
         } catch (Exception e) {
-            throw new IOException("Was not able to fetch object from s3 bucket:" + name);
+            throw new IOException("Was not able to fetch object from bucket:" + name, e);
         }
     }
 
     @Override
     public void deleteAll() throws IOException {
         String objectPrefix = getObjectPrefix();
-        LOG.info("delete all job storage parts from prefix:{}", objectPrefix);
+        LOG.info("delete all job storage parts from prefix: {}", objectPrefix);
 
-        ObjectListing listing = client.listObjects(new ListObjectsRequest().withBucketName(bucketName).withPrefix(objectPrefix));
         try {
+
+            ObjectListing listing = client.listObjects(new ListObjectsRequest().withBucketName(bucketName).withPrefix(objectPrefix));
+
             while (listing != null) {
+
                 List<DeleteObjectsRequest.KeyVersion> keys = new ArrayList<>(listing.getObjectSummaries().size());
+
                 for (S3ObjectSummary summary : listing.getObjectSummaries()) {
                     String key = summary.getKey();
                     LOG.debug("add key to deletion batch, key={}", key);
                     keys.add(new DeleteObjectsRequest.KeyVersion(key));
                 }
+
                 if (!keys.isEmpty()) {
                     LOG.debug("delete key batch");
                     client.deleteObjects(new DeleteObjectsRequest(bucketName).withKeys(keys));
                 }
-                if (!listing.isTruncated())
+
+                if (!listing.isTruncated()) {
                     return;
+                }
 
                 listing = client.listNextBatchOfObjects(listing);
             }
-        } catch (RuntimeException e) {
-            throw new IOException("Cannot delete all parts from prefix:" + objectPrefix, e);
+        } catch (Exception e) {
+            throw new IOException("Cannot delete all parts from prefix: " + objectPrefix, e);
         }
     }
 
@@ -168,23 +182,18 @@ public class AwsS3JobStorage implements JobStorage {
         try {
             return client.doesObjectExist(bucketName, getObjectName(name));
         } catch (RuntimeException e) {
-            throw new IOException("Cannot check existence", e);
+            throw new IOException("Cannot check existence of: " + name, e);
         }
     }
 
-    @Override
-    public void store(String name, InputStream inputStream) throws IOException {
-        requireNonNull(name, "name may not be null!");
-        requireNonNull(inputStream, "inputStream may not be null!");
-
-        storeInS3(name, inputStream, null);
-    }
-
     private void storeInS3(String name, InputStream inputStream, Long contentLength) throws IOException {
+
         try (InputStream stream = inputStream) {
+
             if (!client.doesBucketExistV2(bucketName)) {
                 client.createBucket(bucketName);
             }
+
             ObjectMetadata meta = new ObjectMetadata();
 
             if (contentLength != null) {
@@ -194,11 +203,14 @@ public class AwsS3JobStorage implements JobStorage {
             String objectName = getObjectName(name);
             LOG.debug("Start upload of objectName={} to bucket {}", objectName, bucketName);
 
-            // The Transfer Manager is by default non-blocking, which means it will not wait
-            // for the upload to finish and returns immediately
+            /*
+             * The Transfer Manager is by default non-blocking, which means it will not wait
+             * for the upload to finish and returns immediately - we must wait for
+             * completion later!
+             */
             Upload upload = transferManager.upload(bucketName, objectName, stream, meta);
 
-            ProgressListener progressListener = createProgressListener(upload, name);
+            ProgressListener progressListener = new LogS3ProgressListener(upload, name);
             upload.addProgressListener(progressListener);
 
             try {
@@ -207,73 +219,89 @@ public class AwsS3JobStorage implements JobStorage {
             } catch (AmazonClientException e) {
                 LOG.error("Error while uploading objectName={} to bucket {}", objectName, bucketName);
                 throw e;
+            } finally {
+                /* cleanup */
+                upload.removeProgressListener(progressListener);
             }
 
         } catch (Exception e) {
-            throw new IOException("Store of " + name + " to s3 failed", e);
+            throw new IOException("Store of: " + name + " to S3 bucket: " + bucketName + " failed", e);
         }
     }
 
-    private ProgressListener createProgressListener(Upload upload, String objectName) {
-        ProgressListener progressListener = new ProgressListener() {
-            private int previousPercentTransferred = 0;
+    private String getObjectName(String name) {
+        return getObjectPrefix() + name;
+    }
 
-            private int getPreviousPercentTransferred() {
-                return previousPercentTransferred;
+    private String getObjectPrefix() {
+        return storagePath + "/" + jobUUID + "/";
+    }
+
+    private class LogS3ProgressListener implements ProgressListener {
+
+        private Upload upload;
+        private String objectName;
+
+        private LogS3ProgressListener(Upload upload, String objectName) {
+            this.upload = upload;
+            this.objectName = objectName;
+        }
+
+        private int previousPercentTransferred = 0;
+
+        private int getPreviousPercentTransferred() {
+            return previousPercentTransferred;
+        }
+
+        private void setPreviousPercentTransferred(int previousPercentTransferred) {
+            this.previousPercentTransferred = previousPercentTransferred;
+        }
+
+        private void resetPreviousPercentTransferred() {
+            setPreviousPercentTransferred(0);
+        }
+
+        @Override
+        public void progressChanged(ProgressEvent progressEvent) {
+            if (upload == null) {
+                return;
             }
 
-            private void setPreviousPercentTransferred(int previousPercentTransferred) {
-                this.previousPercentTransferred = previousPercentTransferred;
+            ProgressEventType progressEventType = progressEvent.getEventType();
+
+            if (progressEventType.isByteCountEvent()) {
+                int percentTransferred = (int) upload.getProgress().getPercentTransferred();
+                printStatusMessage(percentTransferred);
             }
 
-            private void resetPreviousPercentTransferred() {
-                setPreviousPercentTransferred(0);
-            }
+            switch (progressEvent.getEventType()) {
+            case TRANSFER_COMPLETED_EVENT:
+                LOG.info("Transfer of {} completed.", objectName);
+                resetPreviousPercentTransferred();
+                break;
+            case TRANSFER_FAILED_EVENT:
+                resetPreviousPercentTransferred();
+                LOG.error("Transfer of {} failed.", objectName);
+                break;
+            case TRANSFER_STARTED_EVENT:
+                LOG.info("Transfer of {} started.", objectName);
+                resetPreviousPercentTransferred();
+                break;
+            default:
+                break;
 
-            @Override
-            public void progressChanged(ProgressEvent progressEvent) {
-                if (upload == null) {
-                    return;
+            }
+        }
+
+        private void printStatusMessage(int percentTransferred) {
+            if (percentTransferred > getPreviousPercentTransferred()) {
+                // print progress only every 10 percent
+                if ((percentTransferred % 10) == 0) {
+                    setPreviousPercentTransferred(percentTransferred);
+                    LOG.debug("Percent transfered: " + percentTransferred);
                 }
-
-                ProgressEventType progressEventType = progressEvent.getEventType();
-
-                if (progressEventType.isByteCountEvent()) {
-                    int percentTransferred = (int) upload.getProgress().getPercentTransferred();
-                    printStatusMessage(percentTransferred);
-                }
-
-                switch (progressEvent.getEventType()) {
-                case TRANSFER_COMPLETED_EVENT:
-                    LOG.info("Transfer of {} completed.", objectName);
-                    resetPreviousPercentTransferred();
-                    break;
-                case TRANSFER_FAILED_EVENT:
-                    resetPreviousPercentTransferred();
-                    LOG.error("Transfer of {} failed.", objectName);
-                    break;
-                case TRANSFER_STARTED_EVENT:
-                    LOG.info("Transfer of {} started.", objectName);
-                    resetPreviousPercentTransferred();
-                    break;
-                default:
-                    break;
-
-                }
             }
+        }
 
-            private void printStatusMessage(int percentTransferred) {
-                if (percentTransferred > getPreviousPercentTransferred()) {
-                    // print progress only every 10 percent
-                    if ((percentTransferred % 10) == 0) {
-                        setPreviousPercentTransferred(percentTransferred);
-                        LOG.debug("Percent transfered: " + percentTransferred);
-                    }
-                }
-            }
-
-        };
-
-        return progressListener;
     }
 }

--- a/sechub-storage-sharedvolume-spring/src/main/java/com/mercedesbenz/sechub/storage/sharevolume/spring/SharedVolumeJobStorage.java
+++ b/sechub-storage-sharedvolume-spring/src/main/java/com/mercedesbenz/sechub/storage/sharevolume/spring/SharedVolumeJobStorage.java
@@ -16,7 +16,6 @@ import java.util.UUID;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.util.FileSystemUtils;
-import org.springframework.web.client.ResourceAccessException;
 
 import com.mercedesbenz.sechub.storage.core.JobStorage;
 import com.mercedesbenz.sechub.storage.core.StorageException;
@@ -41,11 +40,20 @@ public class SharedVolumeJobStorage implements JobStorage {
 
     @Override
     public InputStream fetch(String name) throws IOException {
-        Path path = getPathToFile(name);
-        if (path == null) {
-            return null;
+        try {
+            Path path = getPathToFile(name);
+            if (path == null) {
+                return null;
+            }
+            return new FileInputStream(path.toFile());
+        } catch (Exception e) {
+            throw new IOException("Was not able to fetch: " + name, e);
         }
-        return new FileInputStream(path.toFile());
+    }
+
+    @Override
+    public void store(String name, InputStream stream) throws IOException {
+        store(name, stream, -1);
     }
 
     @Override
@@ -55,14 +63,15 @@ public class SharedVolumeJobStorage implements JobStorage {
 
         if (name.contains("..")) {
             // This is a security check
-            throw new StorageException("Cannot store file with relative path outside current directory " + name);
+            throw new StorageException("Cannot store file with relative path outside current directory: " + name);
         }
 
         try {
             Files.createDirectories(volumePath);
         } catch (IOException e) {
-            throw new StorageException("Could not initialize storage directory at:" + volumePath, e);
+            throw new StorageException("Could not initialize storage directory at: " + volumePath, e);
         }
+
         LOG.info("job:{}: storing {} in path {}", jobUUID, name, storagePath);
 
         Path pathToFile = getPathToFile(name);
@@ -71,22 +80,42 @@ public class SharedVolumeJobStorage implements JobStorage {
             Files.copy(inputStream, pathToFile, StandardCopyOption.REPLACE_EXISTING);
 
             LOG.debug("Stored:{} at {}", name, pathToFile);
-        } catch (ResourceAccessException e) {
-            throw new IOException("Was not able to store input stream into file " + pathToFile, e);
+        } catch (Exception e) {
+            throw new IOException("Was not able to store input stream into file: " + pathToFile, e);
         }
-    }
-
-    private Path getPathToFile(String fileName) {
-        requireNonNull(fileName, "fileName may not be null!");
-        return this.volumePath.resolve(fileName);
     }
 
     public void deleteAll() throws IOException {
-        if (Files.notExists(volumePath)) {
-            return;
+        try {
+            if (Files.notExists(volumePath)) {
+                return;
+            }
+            FileSystemUtils.deleteRecursively(volumePath);
+
+            LOG.info("deleted all inside {}", volumePath);
+
+        } catch (Exception e) {
+            throw new IOException("Was not able to delete all from: " + volumePath, e);
         }
-        FileSystemUtils.deleteRecursively(volumePath);
-        LOG.info("deleted all inside {}", volumePath);
+    }
+
+    public boolean isExisting(String fileName) {
+        return getPathToFile(fileName).toFile().exists();
+    }
+
+    @Override
+    public Set<String> listNames() throws IOException {
+        try {
+            Set<String> names = new LinkedHashSet<>();
+            if (Files.notExists(volumePath)) {
+                Files.createDirectories(volumePath);
+            }
+            Files.list(volumePath).forEach(child -> names.add(child.getFileName().toString()));
+
+            return names;
+        } catch (Exception e) {
+            throw new IOException("Was not able to list names for: " + volumePath, e);
+        }
     }
 
     @Override
@@ -94,25 +123,8 @@ public class SharedVolumeJobStorage implements JobStorage {
         return "SharedVolumeJobStorage [projectId=" + storagePath + ", jobUUID=" + jobUUID + ", path=" + volumePath + "]";
     }
 
-    public boolean isExisting(String fileName) {
-        return getPathToFile(fileName).toFile().exists();
-
+    private Path getPathToFile(String fileName) {
+        requireNonNull(fileName, "fileName may not be null!");
+        return this.volumePath.resolve(fileName);
     }
-
-    @Override
-    public Set<String> listNames() throws IOException {
-        Set<String> names = new LinkedHashSet<>();
-        if (Files.notExists(volumePath)) {
-            Files.createDirectories(volumePath);
-        }
-        Files.list(volumePath).forEach(child -> names.add(child.getFileName().toString()));
-
-        return names;
-    }
-
-    @Override
-    public void store(String name, InputStream stream) throws IOException {
-        store(name, stream, -1);
-    }
-
 }

